crypto/kdf/tls1_prf.c

```diff
@@ -31,145 +32,127 @@ typedef struct {
     /* Buffer of concatenated seed data */
     unsigned char seed[TLS1_PRF_MAXBUF];
     size_t seedlen;
-} TLS1_PRF_PKEY_CTX;
+} TLS1_PRF_KDF_CTX;
```

Comment:
I'm a little worried about this change. The QuickAssist Engine that I maintain provides an implementation of PRF for offloading to hardware. As part of this we need to maintain our own ctx info related to the request.
Currently with the old EVP_PKEY_CTX when we provide our own init, cleanup, ctrl and derive functions we can specify our own internal data structure and assign it to the ctx->data structure that was a void pointer and include our own data structures. As we provide our own init and cleanup functions we had full control as long as we provided our own functions for everything that uses the data structure. With this change we now have a tightly defined TLS1_PRF_KDF_CTX data structure with no field for user defined data. Can we at least have some kind of pointer (perhaps names custom) in the structure that is a void pointer that engines can use to attach their own data to. The same also applies to the HKDF_KDF_CTX and any other KDF CTX structures.

Reply:
Actually am I misunderstand the implementation, as a void pointer is always passed around, does that mean I can define my own data structure within my engine implementation and always cast the ctx to my data structure? If so I take back what I said, the implementation is fine.

Reply:
I'm not sure I understand everything you said there, but it basically sounds like you had your own EVP_PKEY_METHOD.  With this new API, you will have to provide a EVP_KDF_METHOD instead.

Reply:
This is a work-in-progress. Currently there is no way for an engine or an application to provide an EVP_KDF_METHOD. At some point there will be a way to do that; whether it be via engine or some other mechanism is not yet clear.